README.md
# validx

[![npm](https://img.shields.io/npm/v/validx.svg?maxAge=1000)](https://www.npmjs.com/package/validx)
[![dependency Status](https://img.shields.io/david/jeffijoe/validx.svg?maxAge=1000)](https://david-dm.org/jeffijoe/validx)
[![devDependency Status](https://img.shields.io/david/dev/jeffijoe/validx.svg?maxAge=1000)](https://david-dm.org/jeffijoe/validx)
[![Build Status](https://img.shields.io/travis/jeffijoe/validx.svg?maxAge=1000)](https://travis-ci.org/jeffijoe/validx)
[![Coveralls](https://img.shields.io/coveralls/jeffijoe/validx.svg?maxAge=1000)](https://coveralls.io/github/jeffijoe/validx)
[![npm](https://img.shields.io/npm/dt/validx.svg?maxAge=1000)](https://www.npmjs.com/package/validx)
[![npm](https://img.shields.io/npm/l/validx.svg?maxAge=1000)](https://github.com/jeffijoe/validx/blob/master/LICENSE.md)
[![node](https://img.shields.io/node/v/validx.svg?maxAge=1000)](https://www.npmjs.com/package/validx)

Validation library for [MobX](https://github.com/mobxjs/mobx).

## Install

```
# Bring your own MobX
npm install --save mobx validx
```

Table of Contents
=================

   * [Why?](#why)
   * [Examples](#examples)
   * [API documentation](#api-documentation)
      * [The ValidationContext](#the-validationcontext)
         * [`validate()`](#validate)
         * [`reset()`](#reset)
         * [`addErrors()`](#adderrors)
         * [`getErrors()`](#geterrors)
         * [`getError()`](#geterror)
         * [`clearErrors()`](#clearerrors)
         * [`errors`](#errors)
         * [`isValid`](#isvalid)
      * [Validators](#validators)
      * [Built-in validators](#built-in-validators)
         * [`required`](#required)
         * [`pattern`](#pattern)
         * [`func`](#func)
      * [Bound `ValidationContext`](#bound-validationcontext)
   * [Author](#author)

# Why?

We want to reactively display validation issues in our UI when using MobX. You can use
any (non-DOM) validation library, but then you still need to make it reactive.

**ValidX** is built for [MobX](https://github.com/mobxjs/mobx) and is easy to use, yet
powerful enough to add any validation you'd like.

# Examples

See the [TypeScript example][ts-example] and [Babel example][babel-example] for runnable examples (in Node).

# API documentation

## The `ValidationContext`

The meat of the library is the validation context. To create one, you can either use

```js
import { ValidationContext } from 'validx'
const validation = new ValidationContext()
```

Or use the factory function

```js
import { validationContext } from 'validx'
const validation = validationContext()
```

The properties on the context are reactive (observables/computeds powered by MobX).

### `validate()`

Given an object and a schema, validates the object and returns itself.
`validate()` is a MobX `action`.

Calling it will populate the `errors` property with any validation errors
found on the object

There are multiple ways to use `validate()`, see [Bound ValidationContext](#bound-validationcontext).

```js
import { validationContext, required, pattern } from 'validx'

const validation = validationContext()

// For good measure, reset the internal state.
validation.reset()

// validate takes 2 parameters:
// - the object to validate
// - the schema.
const schema = {
  // The schema maps to the fields we are validating.
  name: [
    // Each field you wish to validate is an array of validation
    // functions.
    // `required` is just a function returning a validator function.
    required({ msg: 'Name is required' })
  ],
  email: [
    required({ msg: 'Email is required' }),
    // The `pattern` validator can be used to validate
    // emails and other regexps.
    pattern({ pattern: 'email', msg: 'Not a valid email' })
  ]
}

validation.validate({
  name: '',
  email: ''
}, schema)
```


Now that we have validated our object, we can pull the errors
from the context.

```js
console.log(validation.isValid)
// false

console.log(validation.errors.name)
// ['Name is required']

console.log(validation.errors.email)
// ['Email is required', 'Not a valid email']
```

To validate again, we need to reset the context and then call validate.

```js
validation.reset().validate({
  name: 'Jeff',
  email: 'test'
}, schema)

console.log(validation.isValid)
// false

console.log(validation.errors.name)
// []

console.log(validation.errors.email)
// ['Not a valid email']
```

Let's see what the errors are like when we
log them after resetting.

```js
validation.reset()
console.log(validation.isValid)
// true
console.log(validation.errors.name)
// undefined

console.log(validation.errors.email)
// undefined
```

They are undefined because we don't know
what fields will be validated yet.

```js
validation.validate({
  name: 'Jeff',
  email: 'test'
}, schema)

console.log(validation.isValid)
// false

console.log(validation.errors.name)
// []

console.log(validation.errors.email)
// ['Not a valid email']
```

### `reset()`

Resets the internal state of the context. You usually use this before
starting a new validation. `reset()` is a MobX `action`.

```js
const validation = validationContext()
validation.validate({ name: '' }, {
  name: [required()]
})
console.log(validation.errors.name)
// ['This field is invalid']
validation.reset()
console.log(validation.errors.name)
// undefined
```

**Use case:** sharing a validation context in a class hierarchy.

```js
class Element {
  @observable label = ''

  validation = validationContext(this)

  @action validate () {
    this.validation.validate({
      label: [required({ msg: 'Label required' })]
    })
  }
}

class TextElement extends Element {
  @observable placeholder = ''

  @action validate () {
    // reset before caling super
    this.validation.reset().validate({
      placeholder: [required({ msg: 'Placeholder required' })]
    })
    super.validate()
  }
}

const textElement = new TextElement()
textElement.validate()
console.log(textElement.validation.errors)
// { label: ['Label required'], placeholder: ['Placeholder required'] }

textElement.label = 'First name'
textElement.validate()
console.log(textElement.validation.errors)
// { placeholder: ['Placeholder required'] }
```

### `addErrors()`

If you at some point want to add errors without calling validate, this
is how to do it. `addErrors()` is a MobX `action`.

```js
const obj = {}
const validation = validationContext(obj, {
})

if (usernameIsTaken) {
  validation.addErrors({
    username: ['Username is taken']
  })
}

console.log(validation.errors.username)
// ['Username is taken']
```

### `getErrors()`

Safer way to get errors for a field rather than using `errors.field`,
as this will return an empty array in case there are no errors.

```js
const validation = validationContext()
validation.addErrors({ name: ['Not cool'] })

validation.getErrors('name')
// ['Not cool']

validation.reset()
validation.getErrors('name')
// []
```

### `getError()`

Convenience method for `getErrors('field')[0]`.
Returns `undefined` if the error is not found.

```js
const validation = validationContext()
validation.addErrors({ name: ['Not cool'] })

validation.getError('name')
// 'Not cool'

validation.reset()
validation.getError('name')
// undefined
```

### `clearErrors()`

Clear errors for a single field, instead of reseting the whole context.

```js
const validation = validationContext()
validation.addErrors({ name: ['Not cool'], language: ['Must be JS'] })

validation.getError('name')
// 'Not cool'

validation.clearErrors('name')

validation.getError('name')
// 'Not cool'

validation.getError('language')
// 'Must be JS'
```

### `errors`

A MobX `computed` map of field -> errors. When `validate` discovers validation errors, it
puts them in here.

If a field has not been validated, or the context has just been reset, there
will be no keys.

```js
const validation = validationContext(obj, {
  name: [required()]
})
console.log(validation.errors)
// {}

validation.validate()
console.log(validation.errors)
// { name: ['This field is invalid'] }
validation.validate()
// { name: ['This field is invalid', 'This field is invalid] }
// that's why you need to remember to `reset()` first!
```

### `isValid`

A MobX `computed` property that determines whether the context
has any errors or not.

```js
const validation = validationContext()
console.log(validation.isValid)
// true
validation.validate({ }, { name: [required()] })
console.log(validation.isValid)
// false
```

## Validators

Validators are just plain functions that take an object
with the rule configuration as a parameter and returns
either true for success, false for failure, or a string
for a failure with a specific message.

```js
const validation = validationContext()
const schema = {
  age: [
    required({ msg: 'Age is required' }),
    (opts) => opts.value >= 13 || 'Must be 13 years or older'
  ]
}

validation.validate({
  age: 8
}, schema)

console.log(validation.errors.age)
// ['Must be 13 years or older']
```

## Built-in validators

### `required`

Takes a string (error message) or an object in the form of `{ required: boolean, msg: string }`.

If `required === false`, the validator will always return true.

```js
{
  // default message
  firstName: [required()],
  // shorthand
  lastName: [required('Last name, please?')],
  // Fully-fledged, only enabled if `this.getSmsNotifications` is `true`
  phoneNumber: [
    required({
      msg: 'Please enter your number so we can send you those notifications',
      required: this.getSmsNotifications,
    })
  ]
}
```

### `pattern`

Let's you specify a named pattern or your own regex. Supports simple and config signatures.

Named patterns:

* `'email'` (uses `email-validator`)
* `'url'` (uses `is-url`)

Example:

```js
{
  // simple
  email: [pattern('email')],
  // simple with message
  homepage: [pattern('url', 'Not a valid URL, come on!')],
  // simple regex with message
  bio: [pattern(/javascript/i, 'Sorry - no JS, no job.')],
  // config with regex and message
  twitterHandle: [
    pattern({
      pattern: /^@/,
      msg: 'Twitter handles start with an @'
    })
  ]
}
```

### `func`

If you dislike the `boolean || 'message'` approach to custom validators, you can use `func`.

Example:

```js
{
  knownLibraries: [
    // Without func.
    (opts) => opts.value.includes('react') || 'Sorry, only React devs here!',
    // With func.
    func(
      (opts) => opts.value.includes('react'),
      'Sorry, only React devs here!'
    )
  ]
}
```

## Bound `ValidationContext`

By passing either 1 or 2 parameters to `validationContext`, you can "pre-bind" it
to an object and a schema.

```js
// Not bound to anything.
const obj = { name: '' }
const validation = validationContext()

validation.validate(obj, {
  name: [required()]
})
```

```js
// Bound to an object
const obj = { name: '' }
const validation = validationContext(obj)

validation.validate({
  name: [required()]
})
```

```js
// Bound to an object and a schema
const obj = { name: '' }
const validation = validationContext(obj, {
  name: [required()]
})

validation.validate()
```

# Author

Jeff Hansen - [@Jeffijoe](https://twitter.com/Jeffijoe)

[ts-example]: /examples/typescript
[babel-example]: /examples/babel


CHANGELOG.md
# Changelog

## 1.0.0-alpha.2

* Upgraded to MobX 6
* Removed `peerDependencies` in favor of BYOM (Bring Your Own MobX)
* Overhaul build, lint and test infra

## 1.0.0

* Upgraded to MobX v4, breaks compatibility with MobX <= 3

## 0.2.0

* Added `clearErrors(field: string)` to the context, lets you remove errors for a particular field.

## 0.1.4

* Added `url` pattern validator.
* Added `func` validator
* More documentation

## 0.1.3

* Skip falsy values in schema.

## 0.1.2

* Moved mobx to peerDependencies.

## 0.1.1

* Updated npm description.

## 0.1.0

* Added `validation.getErrors()`.
* Added `validation.getError()`.
* Removed `lodash` dependency.

## 0.0.4

* First official release.


test/index.ts
import 'mocha'

const chai = require('chai')
chai.use(require('sinon-chai'))


test/utils.spec.ts
import { forEach, every, mapToObject } from '../src/utils'
import { spy } from 'sinon'
import { expect } from 'chai'

describe('utils', () => {
  describe('forEach', () => {
    it('calls the iteratee on arrays', () => {
      const arr = ['a', 'b', 'c']
      const iteratee = spy()
      forEach(arr, iteratee)
      expect(iteratee.callCount).to.equal(3)
      expect(iteratee).has.been.calledWith('a', 0, arr)
      expect(iteratee).has.been.calledWith('b', 1, arr)
      expect(iteratee).has.been.calledWith('c', 2, arr)
    })

    it('calls the iteratee on objects', () => {
      const obj = { a: 1, b: 2, c: 3 }
      const iteratee = spy()
      forEach(obj, iteratee)
      expect(iteratee.callCount).to.equal(3)
      expect(iteratee).has.been.calledWith(1, 'a', obj)
      expect(iteratee).has.been.calledWith(2, 'b', obj)
      expect(iteratee).has.been.calledWith(3, 'c', obj)
    })

    it('only iterates own props', () => {
      class Test {
        test: string
        method() {}
      }

      const instance = new Test()
      instance.test = 'hello'
      const iteratee = spy()
      forEach(instance, iteratee)
      expect(iteratee).has.been.calledOnce
      expect(iteratee).has.been.calledWith('hello', 'test', instance)
    })
  })

  describe('every', function () {
    it('returns true if every element satisfies the predicate', () => {
      const arr = [1, 1, 1]
      expect(every(arr, (x) => x === 1)).to.equal(true)
    })

    it('returns false if some element does not satisfy the predicate', () => {
      const arr = [1, 1, 2]
      expect(every(arr, (x) => x === 1)).to.equal(false)
    })

    it('stops early if it encounters false', () => {
      const arr = [1, 2, 1]
      const predicate = spy((x) => x === 1)
      expect(every(arr, predicate)).to.equal(false)
      expect(predicate).has.been.calledTwice
      expect(predicate).has.been.calledWith(1, 0, arr)
      expect(predicate).has.been.calledWith(2, 1, arr)
    })

    it('works on objects', () => {
      expect(every({ a: 1, b: 2 }, (v) => v === 2)).to.equal(false)
      expect(every({ a: 2, b: 2 }, (v) => v === 2)).to.equal(true)
    })
  })

  describe('mapToObject', () => {
    it('maps a map to an object', () => {
      const map = new Map([
        ['one', 1],
        ['two', 2],
      ])
      const obj = mapToObject(map)

      expect(obj['one']).to.equal(1)
      expect(obj['two']).to.equal(2)
    })
  })
})


test/validation.spec.ts
import 'mocha'
import { ValidationContext, IValidationSchema, required, pattern } from '../src'
import { reaction } from 'mobx'
import { expect } from 'chai'
import { spy } from 'sinon'
import { IValidatorOptions, validationContext } from '../src/validation'

describe('ValidationContext', () => {
  describe('#validate', () => {
    it('runs validations on the input object', () => {
      const v = new ValidationContext()
      const reactionSpy = spy()
      reaction(
        () => v.errors,
        (e: any) => {
          expect(Object.keys(e).length).to.equal(2)
          reactionSpy()
        }
      )
      v.validate(
        {
          name: null,
          lol: 'test'
        },
        {
          name: [required({ msg: 'yo' })],
          lol: [required(), pattern({ pattern: 'email' })]
        }
      )

      expect(v.errors['name'].length).to.equal(1)
      expect(v.errors['name'][0]).to.equal('yo')
      expect(v.errors['lol'].length).to.equal(1)
      expect(v.errors['lol'][0]).to.match(/email/i)
      expect(reactionSpy.calledOnce).to.equal(true)
    })

    it('accumulates errors', () => {
      const v = new ValidationContext()
      v.validate(
        {
          name: ''
        },
        {
          name: [required()]
        }
      )
      v.validate(
        {
          email: '',
          name: ''
        },
        {
          email: [required(), pattern({ pattern: 'email' })],
          name: [required()]
        }
      )

      expect(Object.keys(v.errors).length).to.equal(2)
    })

    it('uses default error message when validator returns false', () => {
      const v = validationContext()
      v.validate(
        {
          name: 'Joe'
        },
        {
          name: [(v: IValidatorOptions<{ name: string }>) => v.value === 'Jeff']
        }
      )

      expect(v.errors['name'][0]).to.match(/invalid/i)
    })

    it('has no errors when all is well', () => {
      const c = new ValidationContext()
      c.validate(
        { name: 'heh' },
        {
          name: [required()]
        }
      )
      expect(Object.keys(c.errors).length).to.equal(
        0,
        'there should be no errors'
      )
    })

    it('skips falsy values in schema', () => {
      const c = new ValidationContext()
      c.validate(
        { name: '' },
        {
          name: [undefined as any, null, false, required()]
        }
      )
      expect(Object.keys(c.errors).length).to.equal(
        1,
        'there should be 1 error'
      )
    })
  })

  describe('#reset', () => {
    it('removes all errors', () => {
      const c = new ValidationContext()
      const reactionSpy = spy()
      c.validate(
        { name: null },
        {
          name: [required()]
        }
      )
      expect(Object.keys(c.errors).length).to.equal(
        1,
        'there should be 1 error'
      )
      reaction(
        () => c.errors,
        (e: any) => {
          expect(Object.keys(e).length).to.equal(0)
          reactionSpy()
        }
      )
      c.reset()
      expect(Object.keys(c.errors).length).to.equal(
        0,
        'there should be no errors after resetting'
      )
      expect(reactionSpy.calledOnce).to.equal(true)
    })
  })

  describe('#isValid', () => {
    it('reacts based on errors', () => {
      const reactionSpy = spy()
      const c = new ValidationContext()
      reaction(() => c.isValid, reactionSpy)
      expect(c.isValid).to.equal(true)
      c.validate(
        { name: null },
        {
          name: [required()]
        }
      )
      expect(c.isValid).to.equal(false)
      c.reset()
      expect(c.isValid).to.equal(true)
      expect(reactionSpy.callCount).to.equal(2, 'should have reacted 2 times')
    })
  })

  describe('#getErrors', () => {
    it('returns the errors for the given field', () => {
      const c = validationContext()
      c.addErrors({ test: ['Hello', 'World'] })
      expect(c.getErrors('test')).to.deep.equal(['Hello', 'World'])
    })

    it('returns an empty array for the given field if there are no errors', () => {
      const c = validationContext()
      expect(c.getErrors('test')).to.deep.equal([])
    })

    it('supports bound contexts', () => {
      const c = validationContext(
        { name: '' },
        {
          name: [required()]
        }
      )
      c.reset().validate()
      expect(c.getErrors('name')[0]).to.match(/required/)
    })
  })

  describe('#getError', () => {
    it('returns the first error for the given field', () => {
      const c = validationContext()
      c.addErrors({ test: ['Hello', 'World'] })
      expect(c.getError('test')).to.deep.equal('Hello')
    })

    it('returns an empty array for the given field if there are no errors', () => {
      const c = validationContext()
      expect(c.getError('test')).to.deep.equal(undefined)
    })

    it('supports bound contexts', () => {
      const c = validationContext(
        { name: '' },
        {
          name: [required('hah'), pattern({ pattern: 'email' })]
        }
      )
      c.reset().validate()
      expect(c.getError('name')).to.equal('hah')
    })
  })

  describe('#clearErrors', function() {
    it('clears errors for the specified field', () => {
      const c = validationContext()
      c.addErrors({ test: ['Hello'] })
      expect(c.getError('test')).to.deep.equal('Hello')
      c.clearErrors('test')
      expect(c.getError('test')).to.be.undefined
    })
  })

  describe('validators', () => {
    describe('regular function', () => {
      it('works', () => {
        type ITest = { name: string; age: number }
        const schema: IValidationSchema<ITest> = {
          name: [
            opts => {
              return opts.obj.name === 'Jeff' ? true : 'jeff plz'
            }
          ]
        }
        const c = new ValidationContext().validate<ITest>(
          {
            name: 'Jeff',
            age: 22
          },
          schema
        )
        expect(c.isValid).to.equal(true)
        c.validate(
          {
            name: 'Joe',
            age: 24
          },
          schema
        )
        expect(c.isValid).to.equal(false)
        expect(c.errors['name'][0]).to.equal('jeff plz')
      })
    })
  })

  describe('#addErrors', () => {
    it('adds errors to the context', () => {
      const c = new ValidationContext()
      expect(c.isValid).to.equal(true)
      c.addErrors({
        test: ['yeah']
      })

      expect(c.errors['test'][0]).to.equal('yeah')
      expect(c.isValid).to.equal(false)
    })
  })
})

describe('validationContext(object)', () => {
  describe('when called with no parameter', () => {
    it('returns a regular validation context', () => {
      const v = validationContext()
      v.validate(
        { name: '' },
        {
          name: [required()]
        }
      )
      expect(v.errors['name'].length).to.equal(1)
    })
  })

  describe('when called with an object parameter', () => {
    it('returns a bound validation context', () => {
      const o = { name: '' }
      const v = validationContext(o)
      v.validate({
        name: [required()]
      })
      expect(v.errors['name'].length).to.equal(1)
      v.reset()
      o.name = 'joe'
      v.validate({
        name: [required()]
      })
      expect(Object.keys(v.errors).length).to.equal(0)
    })
  })

  describe('when called with an object + schema', () => {
    it('returns a bound validation context with a schema', () => {
      const o = { name: '' }
      const v = validationContext(o, {
        name: [required()]
      })
      v.validate()
      expect(v.errors['name'].length).to.equal(1)
      v.reset()
      o.name = 'joe'
      v.validate()
      expect(Object.keys(v.errors).length).to.equal(0)
    })
  })
})


src/utils.ts
export type Iteratee<V, K, O, R> = (value: V, key: K, source: O) => R

/**
 * A specialized version of `forEach` for arrays.
 *
 * From lodash.
 */
function arrayEach<T>(
  array: T[],
  iteratee: Iteratee<T, number, T[], any>
): T[] {
  let index = -1
  const length = array.length

  while (++index < length) {
    if (iteratee(array[index], index, array) === false) {
      break
    }
  }
  return array
}

/**
 * A specialized version of `forEach` for objects.
 */
function objectEach<T extends Record<string, unknown>>(
  source: T,
  iteratee: Iteratee<any, string, T, any>
): T {
  for (let key in source) {
    /* istanbul ignore next */
    if (source.hasOwnProperty(key)) {
      if (iteratee(source[key], key, source) === false) {
        break
      }
    }
  }

  return source
}

/**
 * Invokes the iteratee for each item in the array or object.
 */
export function forEach<T>(
  array: T[],
  iteratee: Iteratee<T, number, T[], any>
): T[]
export function forEach<T>(
  source: T,
  iteratee: Iteratee<any, string, T, any>
): T
export function forEach(
  source: any,
  iteratee: Iteratee<any, any, any, any>
): any {
  const func: any = Array.isArray(source) ? arrayEach : objectEach
  return func(source, iteratee)
}

/**
 * Determines if every element in the collection statisfies the predicate.
 * @type {T[]}
 */
export function every<T>(
  array: T[],
  predicate: Iteratee<T, number, T[], boolean>
): boolean
export function every<T>(
  source: T,
  predicate: Iteratee<any, string, T, boolean>
): boolean
export function every(
  source: any,
  predicate: Iteratee<any, any, any, any>
): any {
  let result = true
  forEach(source, (value, key, source) => {
    if (!predicate(value, key, source)) {
      result = false
      return false
    }
  })
  return result
}

/**
 * Creates an object hash from a map.
 *
 * @param map
 */
export function mapToObject<K extends string, V>(map: Map<K, V>): Record<K, V> {
  const result: Record<K, V> = {} as any
  const entries = Array.from(map.entries())
  for (let index = 0; index < entries.length; index++) {
    const [key, value] = entries[index]
    result[key] = value
  }

  return result
}


src/validation.ts
import {
  observable,
  extendObservable,
  action,
  computed,
  ObservableMap,
  toJS,
} from 'mobx'
import { forEach, every, mapToObject } from './utils'

/**
 * A validatable object where all keys are strings.
 */
export type Validatable = Record<string, unknown>

/**
 * Options passed to validators.
 *
 * @export
 * @interface IValidatorOptions
 * @template T The type of the object being validated.
 * @template R The rule type.
 */
export interface IValidatorOptions<T> {
  field: string
  value: any
  obj: T
}

/**
 * Definition of a validator function.
 *
 * @export
 * @interface IValidator
 * @template T Type of the object being validated.
 * @template R The rule type.
 */
export interface IValidator<T> {
  (opts: IValidatorOptions<T>): boolean | string
}

/**
 * The interface all rules must satisfy.
 *
 * @export
 * @interface IRule
 * @template T
 */
export interface IRule {
  msg?: string
}

/**
 * Maps fields on an object to one or more rules.
 *
 * @export
 * @interface IValidationSchema
 * @template T The object type.
 */
export type IValidationSchema<T> = { [P in keyof T]?: Array<IValidator<T>> }

/**
 * Validation errors are stored as a map of fields to error strings.
 *
 * @export
 * @interface IValidationErrors
 */
export interface IValidationErrors {
  [key: string]: Array<string>
}

/**
 * The validation context interface.
 *
 * @export
 * @interface IValidationContext
 */
export interface IValidationContext {
  errors: IValidationErrors
  isValid: boolean
  reset(): this
  validate<T>(obj: T, schema: IValidationSchema<T>): this
  addErrors(errors: IValidationErrors | { [key: string]: string[] }): this
  getErrors(field: string): string[]
  getError(field: string): string | undefined
  clearErrors(field: string): this
}

/**
 * Validation context with the object already bound to the validate function.
 */
export interface IBoundValidationContext<T> {
  errors: IValidationErrors
  isValid: boolean
  reset(): this
  validate(schema: IValidationSchema<T>): this
  addErrors(errors: IValidationErrors | { [key: string]: string[] }): this
  getErrors(field: keyof T): string[]
  getError(field: keyof T): string | undefined
  clearErrors(field: string): this
}

/**
 * Validation context with the object already bound to the validate function.
 */
export interface ISchemaBoundValidationContext<T>
  extends IBoundValidationContext<T> {
  validate(): this
}

/**
 * Implementation of the validation context.
 *
 * @export
 * @class ValidationContext
 * @implements {IValidationContext}
 */
export class ValidationContext implements IValidationContext {
  /**
   * All validation errors are stored here. To clear, call `reset`.
   *
   * @readonly
   * @type {IValidationErrors}
   * @memberOf ValidationContext
   */
  readonly errors!: IValidationErrors

  /**
   * Determines if the validation context is in a valid state (no errors)
   *
   * @readonly
   * @type {boolean}
   * @memberOf ValidationContext
   */
  readonly isValid!: boolean

  /**
   * Internal map of the errors.
   *
   * @private
   *
   * @memberOf ValidationContext
   */
  private errorsMap!: ObservableMap<string, string[]>

  /**
   * Initializes a new instance of ValidationContext.
   */
  constructor() {
    this.reset = action(this.reset)
    this.addErrors = action(this.addErrors)
    this.clearErrors = action(this.clearErrors)
    this.validate = action(this.validate)
    extendObservable(
      this,
      {
        errorsMap: observable.map<string, string[]>(),
        get errors() {
          return mapToObject(toJS(this.errorsMap))
        },
        get isValid() {
          return every(this.errors, (arr: string[]) => arr.length === 0)
        },
      },
      {
        errors: computed,
        isValid: computed,
      }
    )
  }

  /**
   * Resets the errors.
   *
   * @returns {IValidationContext}
   *
   * @memberOf ValidationContext
   */
  reset(): this {
    this.errorsMap.clear()
    return this
  }

  /**
   * Validates the input object and stores any errors that may have occurred
   * in `errors`.
   *
   * @template T The type of the object being validated.
   * @param {T} obj
   * @param {IValidationSchema<T>} schema
   * @returns {IValidationContext}
   *
   * @memberOf ValidationContext
   */
  validate<T>(obj: T, schema: IValidationSchema<T>): this {
    forEach(schema, (validators: Array<IValidator<T>>, field: string) => {
      const errors = this.ensureErrors(field)
      const value = (obj as any)[field]
      forEach(validators, (validator?: IValidator<T>) => {
        if (!validator) {
          return
        }

        const opts: IValidatorOptions<T> = {
          field,
          value,
          obj,
        }

        let msg = 'This field is invalid.'
        const result = validator(opts)
        if (result === true) {
          return
        } else if (result !== false) {
          msg = result
        }

        errors.push(msg)
      })
    })
    this.cleanupErrors()
    return this
  }

  /**
   * Adds errors to the context.
   */
  addErrors(errors: IValidationErrors | { [key: string]: string[] }) {
    forEach(errors, (arr: string[], field: string) => {
      this.ensureErrors(field).push(...arr)
    })
    return this
  }

  /**
   * Gets the errors for the given field.
   */
  getErrors(field: string) {
    const errors = this.errors[field]
    if (!errors) {
      return []
    }

    return errors.slice()
  }

  /**
   * Gets the first error for the given field.
   * If not found, returns undefined.
   */
  getError(field: string) {
    return this.getErrors(field)[0]
  }

  /**
   * Removes errors for a particular field.
   */
  clearErrors(field: string) {
    this.errorsMap.set(field, [])
    return this
  }

  /**
   * Ensures that an entry in the internal error map
   * exists for the specified field.
   *
   * @private
   * @param {string} field
   * @returns
   *
   * @memberOf ValidationContext
   */
  private ensureErrors(field: string) {
    let errors = this.errorsMap.get(field)
    if (!errors) {
      errors = observable.array([])
      this.errorsMap.set(field, errors)
    }

    return errors
  }

  /**
   * At the end of a validation run, if a field
   * has no errors, it's entry in the error map
   * is removed.
   *
   * @private
   *
   * @memberOf ValidationContext
   */
  private cleanupErrors() {
    const entries = Array.from(this.errorsMap.entries()).filter(
      ([, value]) => value.length === 0
    )
    entries.forEach(([key]) => this.errorsMap.delete(key))
  }
}

export function validationContext<T>(
  objectToValidate: T,
  schema: IValidationSchema<T>
): ISchemaBoundValidationContext<T>
export function validationContext<T>(
  objectToValidate: T
): IBoundValidationContext<T>
export function validationContext(): IValidationContext
export function validationContext<T>(
  objectToValidate?: any,
  schema?: IValidationSchema<T>
):
  | IValidationContext
  | IBoundValidationContext<T>
  | ISchemaBoundValidationContext<T> {
  const v = new ValidationContext()
  if (objectToValidate !== null && objectToValidate !== undefined) {
    if (schema) {
      v.validate = v.validate.bind(v, objectToValidate, schema)
    } else {
      v.validate = v.validate.bind(v, objectToValidate)
    }
  }
  return v
}


src/validators/patternValidator.ts
import { IRule, IValidator } from '../validation'
import * as emailValidator from 'email-validator'
import isUrl = require('is-url')

export type Pattern = 'email' | 'url' | RegExp

/**
 * Rule specifics.
 *
 * @export
 * @interface IPatternRule
 * @extends {IRule<any>}
 */
export interface IPatternRule extends IRule {
  pattern: Pattern
}

const DEFAULT_MESSAGE = 'This field is invalid'

const PatternMessages = {
  email: 'This is not a valid email',
  url: 'This is not a valid url',
}

/**
 * The validator function.
 */
export const pattern = (
  rule: IPatternRule | Pattern,
  msg?: string
): IValidator<any> => {
  const validator: IValidator<any> = (opts) => {
    let valid = false
    const pattern =
      typeof rule === 'string' || rule instanceof RegExp ? rule : rule.pattern
    const message =
      typeof rule === 'string' || rule instanceof RegExp ? msg : rule.msg

    if (pattern === 'email') {
      valid = emailValidator.validate(opts.value)
    } else if (pattern === 'url') {
      valid = isUrl(opts.value)
    } else {
      valid = pattern.test(opts.value)
    }

    return (
      valid ||
      message ||
      (typeof pattern === 'string' && PatternMessages[pattern]) ||
      DEFAULT_MESSAGE
    )
  }

  return validator
}


examples/typescript/src/index.ts
import {
  validationContext,
  required,
  pattern,
  IValidatorOptions,
} from '../../../src'
import { action, autorun } from 'mobx'
import { inspect } from 'util'

interface IObj {
  name?: string
  email?: string
  age?: number
}

// Does not have to be an observable,
// so to demonstrate I'll use a plain object.
const obj: IObj = {}

// First parameter is the object to validate, second
// is the validation schema.
const validation = validationContext(obj, {
  name: [required({ msg: 'Name is required' })],
  email: [
    required('Email is required'), // shorthand message parameter for the required validator.
    pattern({ pattern: 'email', msg: 'That is not a valid email' }),
  ],
  age: [
    required('Age is required'),
    pattern({
      pattern: /\d/, // regex pattern for numbers
      msg: 'Age must be a number',
    }),
    // And a custom validator.
    (opts: IValidatorOptions<IObj>) => {
      // Validators return either true for success, and false (for the default error message)
      // or a string (for a custom error message)
      return opts.value >= 13 || 'You must be over 13 years of age to sign up.'
    },
  ],
})

// Let's set up an autorun that will log whenever
// our validation state changes.
// Keep in mind the first output will say all is well,
// but thats because we didnt start validating yet.
autorun(() => {
  console.log('')
  console.log('')
  console.log('-------- Validation Summary ---------')
  console.log('Bad field count: ', Object.keys(validation.errors).length)
  Object.keys(validation.errors).map((key) => {
    // errors[key] is an observable array.
    // using slice to print them nicely in the console.
    console.log(
      `Errors for ${key}: `,
      // Add some color to the console output. :)
      inspect(validation.errors[key].slice(), { colors: true })
    )
  })
  console.log(
    '-- So is it valid?',
    validation.isValid ? 'Yes it is!' : 'Hell naw!'
  )
})

// The context needs to be reset every time we validate
// from scratch. See the docs as to why this is useful.
// Wrap it in an action so we batch the resetting and validating.
const validate = action(() => validation.reset().validate())

// First round
validation.validate()

// Let's add our name.
obj.name = 'Jeff'

// We need to validate again.
validate()

// Add an email, except it's not really an email..
obj.email = 'test'
validate()

// Much better!
obj.email = 'test@test.com'
validate()

// Alright now the age..
obj.age = 'twentytwo' as any // trick TS
validate()

// Oh, it has to be a number.
obj.age = 2
validate()

// Woops, forgot a digit.
obj.age = 22
validate()

// And we're good!
